terraform/mssql/provision/data.tf

```diff
@@ -1,14 +1,117 @@
 }
 
-data "aws_vpc" "vpc" {
+data "aws_vpc" "default" {
+  count   = length(var.aws_vpc_id) > 0 ? 0 : 1
   default = true
 }
 
-data "aws_subnets" "all" {
+data "aws_subnets" "in_default_vpc" {
+  count = length(var.aws_vpc_id) > 0 ? 0 : 1
   filter {
     name   = "vpc-id"
-    values = [data.aws_vpc.vpc.id]
+    values = [data.aws_vpc.default[0].id]
+  }
+
+  lifecycle {
+    postcondition {
+      condition     = length(self.ids) > 0
+      error_message = "the default vpc doesn't contain any subnets"
+    }
+    postcondition {
+      condition     = length(self.ids) <= 20
+      error_message = "the default vpc contains more than 20 subnets. please specify a different aws_vpc_id or a valid rds_subnet_group containing the desired subnets"
+    }
+  }
+}
+
+data "aws_vpc" "provided" {
+  count = length(var.aws_vpc_id) > 0 ? 1 : 0
+  id    = var.aws_vpc_id
+}
+
+data "aws_subnets" "in_provided_vpc" {
+  count = length(var.aws_vpc_id) > 0 ? 1 : 0
+  filter {
+    name   = "vpc-id"
+    values = [var.aws_vpc_id]
+  }
+
+  lifecycle {
+    postcondition {
+      condition     = length(self.ids) > 0
+      error_message = "the specified aws_vpc_id doesn't contain any subnets"
+    }
+    postcondition {
+      condition     = length(self.ids) <= 20
+      error_message = "the specified aws_vpc_id contains more than 20 subnets. please specify a different aws_vpc_id or a valid rds_subnet_group containing the desired subnets"
+    }
+  }
+}
+
+data "aws_db_subnet_group" "provided" {
+  count = length(var.rds_subnet_group) > 0 ? 1 : 0
+  name  = var.rds_subnet_group
+
+  lifecycle {
+    precondition {
+      condition     = length(var.aws_vpc_id) > 0
+      error_message = "when specifying rds_subnet_group please specify also the corresponding aws_vpc_id"
+    }
+    postcondition {
+      condition     = var.aws_vpc_id == self.vpc_id
+      error_message = "the specified rds_subnet_group doesn't correspond to the specified aws_vpc_id"
+    }
+  }
+}
+
+data "aws_security_groups" "provided" {
+  count = length(var.rds_vpc_security_group_ids) > 0 ? 1 : 0
+  filter {
+    name   = "vpc-id"
+    values = [data.aws_vpc.provided[0].id]
+  }
+  filter {
+    name   = "group-id"
+    values = split(",", var.rds_vpc_security_group_ids)
+  }
+
+  lifecycle {
+    precondition {
+      condition     = length(var.aws_vpc_id) > 0
+      error_message = "when specifying rds_vpc_security_group_ids please specify also the corresponding aws_vpc_id"
+    }
+    postcondition {
+      condition     = length(self.vpc_ids) == 1 && contains(self.vpc_ids, var.aws_vpc_id)
+      error_message = "the specified security groups don't exist or don't correspond to the specified vpc (1)"
+    }
+    postcondition {
+      condition     = split(",", var.rds_vpc_security_group_ids) == self.ids
+      error_message = "the specified security groups don't exist or don't correspond to the specified vpc (2)"
+    }
+  }
+}
+
+data "aws_subnets" "in_provided_subnet_group" {
```

Comment:
newbie question: I am not sure we need this. We already have a `data` resource for the `aws_db_subnet_group` `provided`, which already tell us the subnet ids. Can we just check in that resource that length(self.subnet_ids) > 0 as a postcondition and get rid of this data source?

Reply:
Spot-on. You are absolutely right!
Thanks for simplifying the logic and making it more concise. :)